<html>
	<head>
	<title>Proteinortho Manual / PoFF Manual</title>
	</head>
<body>
<h1>Proteinortho Manual / PoFF Manual</h1>
<small>This manual corresponds to version 5.11</small>
<h2>Introduction</h2>
Proteinortho is a tool to detect orthologous genes within different species.
For doing so, it compares similarities of given gene sequences and clusters them to find significant groups.
The algorithm was designed to handle large-scale data and can be applied to hundreds of species at one.
Details can be found in <a href="http://www.biomedcentral.com/1471-2105/12/124" target="_blank">Lechner <i>et al.</i>, BMC Bioinformatics. 2011 Apr 28;12:124.</a>
<br>
To enhance the prediction accuracy, the relative order of genes (synteny) can be used as additional feature for the discrimination of orthologs.
The corresponding extension, namely PoFF (manuscript in preparation), is already build in Proteinortho.

<h2>Installation</h2>

<h3>Prerequisites</h3>
Proteinortho uses standard software which is often installed already or is part of then package repositories and can thus easily be installed.
To run Proteinortho, you need
<ul>
	<li><a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download">NCBI BLAST+</a>
	    or <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download">NCBI BLAST legacy</a>
	    <br>(to test this, type <code>tblastn</code> (BLAST+) or <code>blastall</code> (BLAST+) in the command line)
	<li><a href="http://www.perl.org/get.html">Perl</a> v5.08 or higher
	    <br>(to test this, type <code>perl -v</code> in the command line)
	<li><a href="https://www.python.org/">Python</a> v2.6.0 or higher to include synteny analysis
	    <br>(to test this, type <code>python -V</code> in the command line)
</ul>

The sources come with a precompiled version of Proteinortho for 64bit Linux. If you want/need to recompile and install it, you will also need
<ul>
	<li><a href="http://www.gnu.org/software/make/">GNU make</a> (to test this, type <code>make</code> in the command line)
	<li><a href="http://gcc.gnu.org">GNU g++</a> (to test this, type <code>g++</code> in the command line)
</ul>

<h3>Building and installing from source</h3>
<ul>
<li> Fetch the latest source code archive from <a href="http://www.bioinf.uni-leipzig.de/Software/proteinortho/" target="_blank">www.bioinf.uni-leipzig.de/Software/proteinortho/</a>.
<li> Extract the files e.g. via <code>tar -xzvf proteinortho_v5.06.tar.gz</code>
<li> Change directory into the extracted folder e.g. via <code>cd proteinortho_v5.06</code>
<li> You can now run <code>./proteinortho5.pl</code> directly
<li> If you want to recompile and install Proteinortho, type <code>make</code> followed <code>sudo make install</code> (requires root privileges).
<li> In any case, run <code>make test</code> to make sure Proteinortho works as expected
</ul>

<h2>Quick Start</h2>
<code>proteinortho5.pl [OPTIONS] FASTA1 FASTA2 [FASTAn...]</code> performs an orthology analysis for the given sets of proteins.
Add <code>-p=blastn</code> in case your sequences are represented as nucleotides (ACTG) rather than as amino acids.

<h2>Quick Tutorial</h2>
This manual assumes that you have Proteinortho installed on your system and can thus directly run it via <code>proteinortho5.pl</code>.
If you have it not installed but only downloaded and extracted to a folder, please use <code>/FULL/PATH/TO/proteinortho5.pl</code> instead.
<p>
Proteinortho assumes, that you have all your gene sequences in <a href="http://en.wikipedia.org/wiki/FASTA_format" target="_blank">FASTA format</a> either
represented as amino acids or as nucleotides. The source code archive contains some examples, namely <i>C.faa</i>, <i>E.faa</i>, <i>L.faa</i>, <i>M.faa</i> located in the <i>test/</i> directory.
By default Proteinortho assumes amino acids and thus uses <i>blastp+</i> to compare sequences. If you have nucleotide sequences, you need to change this by adding the parameter
<i>-p=blastn+</i>. (In case you have only have NCBI BLAST legacy installed, you need to tell this too - either by adding <i>-p=blastp</i> or <i>-p=blastn</i> respectively.)
The full command for the example files would thus be <code>proteinortho5.pl -project=test test/C.faa test/E.faa test/L.faa test/M.faa</code>.
Instead of naming the FASTA files one by one, you could also supply <i>test/*.faa</i> as argument.
Please note that the parameter <i>-project=test</i> is optional. With this, you can set the prefix of the output files generated by Proteinortho.
If you skip the project parameter, the default project name will be <i>myproject</i>.

<p>
Proteinortho will automatically determine the number of available CPU threads and use them accordingly to speed up the calculations. You can use the parameter <i>-cpus=</i> to manually set the number of threads.
When the analysis is done you will find a new file in your current working directory, namely <i>test.proteinortho</i>. To have a quick look, you can i.e. use <code>less -S test.proteinortho</code>.
The tab-separated output generated looks like this:

<pre>
# Species	Genes	Alg.-Conn.	M.faa	L.faa	C.faa		E.faa
4		4	1		M_10	L_10	C_10		E_10
4		4	1		M_11	L_11	C_11		E_11
4		4	1		M_14	L_14	C_14		E_14
...
4		5	0.2		M_19	L_19	C_22,C_63	E_19
...
</pre>
The first line starting with <i>#</i>is a comment line indicating the meaning of each column for each of the following lines which represent an orthologous group each.
The very first column indicates the number of species covered by this group. The second column indicates the number of genes included in the group.
Often, this number will equal the number of species, meaning that there is a single ortholog in each species.
If the number of genes is bigger than the number of species, there are co-orthologs present.
The third column gives rise to the algebraic connectivity of the respective group. Basically, this indicates how densely the genes are connected in the orthology graph that was used for clustering.
A connectivity of 1 indicates a perfect dense cluster with each gene similar to each other gene.
By default, Proteinortho splits each group into two more dense subgroups when the connectivity is below 0.1.
In the second last line of the example above, there is a group with three paralogs in species C (<i>C.faa</i>).
They are separated by a comma (<i>,</i>) indicating that they are co-orthologous the genes in the other species.
<p>
The PoFF extension allows you to use the relative order of genes (synteny) as an additional criterion to disentangle complex co-orthology relations.
To do so, add the parameter <i>-synteny</i>.
You can use it to either come closer to one-to-one orthology relations by preferring synthetically conserved copies in the presence of two very similar paralogs (default),
or just to reduce noise in the predictions by detecting multiple copies of genomic areas (add the parameter <i>-dups=3</i>).
Please note that you need additional data to include synteny, namely the gene positions in <a href="http://www.sequenceontology.org/gff3.shtml">GFF3 format</a>.
As Proteinortho is primarily made for proteins, it will only accept GFF entries of type <i>CDS</i> (column #3 in the GFF-file). 
The attributes column (#9) must contain Name=<i>GENE IDENTIFIER</i> where <i>GENE IDENTIFIER</i> corresponds to the respective identifier in the FASTA format.
It may not contain a semicolon (<i>;</i>)! Alternatively, you can also set ID=<i>GENE IDENTIFIER</i>.
Example files are provided in the source code archive.
Hence, we can run <code>proteinortho5.pl -project=test -synteny test/A1.faa test/B1.faa test/E1.faa test/F1.faa</code> to add synteny information to the calculations.
Of course, this only makes sense if species are sufficiently similar. You won't gain much when comparing e.g. bacteria with fungi.
When the analysis is done you will find an additional file in your current working directory, namely <i>test.poff</i>.
This file is equivalent to the <i>.proteinortho</i> file (above) but can be considered more accurate as synteny was involved for its construction.
<p>
In addition Proteinortho will generate graph files containing all pairwise orthology relationships including similarity scores.
If they are not generated, rerun Proteinortho with the <i>-graph</i> parameter.
<p>
<i>myproject.blast-graph</i>:
filtered raw blast data based on adaptive reciprocal best blast matches (= reciprocal best match plus all reciprocal matches within a range of 95% by default)<br>
<i>myproject.proteinortho-graph</i>:
clustered blast graph. Its connected components are represented in myproject.proteinortho.<br>
<i>myproject.ffadj-graph</i>:
filtered blast data based on adaptive reciprocal best blast matches and synteny (only if <i>-synteny</i> is set) <br>
<i>myproject.poff-graph</i>
clustered ffadj graph. Its connected components are represented in myproject.poff (only if <i>-synteny</i> is set)
<p>
The format of all graph files looks about similar:
<pre>
# file_a        file_b
# a     b       evalue_ab       bitscore_ab     evalue_ba       bitscore_ba
# M.faa 	L.faa
M_15    L_15    0.0     	893     	0.0     	893
M_16    L_16    3e-175  	481     	3e-175  	481
M_19    L_19    8e-93   	262     	8e-93   	262
...
# M.faa E.faa
M_10    E_10    3e-137  	415     	2e-148  	441
M_11    E_11    2e-71   	221     	9e-68   	209
...
</pre>
The first two rows are just comments explaining the meaning of each row.
Whenever a comment line (starting with #) follows, it indicates results comparing the two species is about to follow.
E.g. <i>#M.faa 	L.faa</i> tells that the next lines represent results for species M and L.
All matches are reciprocal matches. If e.g. a match for <i>M_15    L_15</i> is shown,  <i>L_15    M_15</i> exists implicitly.
E-Values and bit scores for both directions are given behind each match.
<p>
The synteny based graph files (<i>myproject.ffadj-graph</i> and <i>myproject.poff-graph</i>) have two additional columns:
<i>same_strand</i> and <i>simscore</i>. The first one indicates if two genes from a match are located at the same strands (1) or not (-1).
The second one is an internal score which can be interpreted as a normalized weight ranging from 0 to 1 based on the respective e-values.
Moreover, a second comment line is followed after the species lines, e.g.
<pre>
...
# M.faa L.faa
# Scores: 4     39      34.000000       39.000000
...
</pre>
These scores are derived from the ffadj algorithm comparing the gene similarities and gene orders in the respective species.
They are:
<ul>
<li>the number of breakpoints to match both gene orders
<li>the number of edges required to match both gene orders
<li>calculated weight of adjacencies
<li>calculated weight of gene similarities
</ul>

<h2>Hints</h2>
<ul>
<li>Using <i>.faa</i> to indicate that your file contains amino acids and <i>.fna</i> to show it contains nucleotides makes life much easier.
<li>Each sequence ID in your FASTA files must be unique. Consider renaming the otherwise.
<li>You need write permissions in the directory of your FASTA files as Proteinortho will create blast databases. If this is not the case, consider using symbolic links to the FASTA files.
<li>The directory <i>tools</i> cotains useful tools, e.g. <code>grab_proteins.pl</code> which fetches protein sequences of orthologous groups from Proteinortho output table
</ul>

<h2>Usage</h2>
<code>proteinortho5.pl [OPTIONS] FASTA1 FASTA2 [FASTAn...]</code> <p>

<table border="1" cellspacing="0" cellpadding="2">
<tr><td><b>Option</b></td><td><b>Default value</b></td><td><b>Description</b></td></tr>

<tr><td>-e=</td><td>1e-05</td><td>E-value for blast</td></tr>
<tr><td>-p=</td><td>blastp+</td><td>blast program to use:
<ul>
<li>blastn+ &#8594; sequences are given as nucleotides
<li>blastp+ &#8594; sequences are given as amino acids
<li>blastn &#8594; sequences are given as nucleotides and blast legacy must be used
<li>blastp &#8594; sequences are given as amino acids and blast legacy must be used
</ul></td></tr>

<tr><td>-project=</td><td>myproject</td><td> prefix for all result file names</td></tr>
<tr><td>-synteny</td><td>-</td><td>activate PoFF extension to separate similar sequences using synteny (requires a GFF file for each FASTA file)</td></tr>
<tr><td>-nograph</td><td>-</td><td>do not generate .graph files with pairwise orthology data<br>saves some time</td></tr>
<tr><td>-keep</td><td>-</td><td>store temporary blast results for reuse</td></tr>
<tr><td>-clean</td><td>-</td><td>remove all unnecessary files after processing</td></tr>
<tr><td>-desc</td><td>-</td><td>write gene description file (XXX.descriptions); works only with NCBI-formated FASTA entries currently</td></tr>
<tr><td>-force</td><td>-</td><td>force recalculation of blast results in any case</td></tr>
<tr><td>-cpus=</td><td>auto</td><td>use the given number of threads</td></tr>
<tr><td>-selfblast</td><td>-</td><td>apply selfblast to directly paralogs; normally these are inferred indirectly from orthology data to other species (experimental!)</td></tr>
<tr><td>-singles</td><td>-</td><td>Also report genes without orthologs in table output</td></tr>
<tr><td>-identity=</td><td>25</td><td>min. percent identity of best blast alignments</td></tr>
<tr><td>-cov=</td><td>50</td><td>min. coverage of best blast alignments in percent</td></tr>
<tr><td>-conn=</td><td>0.1</td><td>min. algebraic connectivity of orthologous groups during clustering</td></tr>
<tr><td>-sim=</td><td>0.95</td><td>min. similarity for additional hits</td></tr>
<tr><td>-blastpath=</td><td>-</td><td>path to your local blast installation (if not in present in default paths)</td></tr>
<tr><td>-blastParameters=</td><td>-</td><td>additional parameters for blast; set these in quotes (e.g. -blastParameters='-seg no')</td></tr>
<tr><td>-verbose</td><td>-</td><td>give a lot of information about the current progress</td></tr>
<tr><td>-debug</td><td>-</td><td>gives detailed information for bug tracking</td></tr>

<tr><td colspan=3><b>Parameters needed to distribute the runs over several machines</b></td></tr>
<tr><td>-step=</td><td>0</td><td>
perform only specific steps of the analysis
<ul>
<li>1 &#8594; generate indices
<li>2 &#8594; perform pairwise analyses
<li>3 &#8594; perform clustering
<li>0 &#8594; perform all steps
</ul></td></tr>
<tr><td>startat=</td><td>0</td><td>file number to start with</td></tr>
<tr><td>stopat=</td><td>-1</td><td>file number to end with (very last one = -1)</td></tr>

<tr><td colspan=3><b>Parameters applied in combination with -synteny<b></td></tr>
<tr><td>-dups=</td><td>0</td><td> applied in combination with -synteny; number of reiterations for adjacencies heuristic to determine duplicated regions; <br>
if set to a higher number, co-orthologs will tend to get clustered together rather than getting separated
</td></tr>
<tr><td>-cs=</td><td>3</td><td> applied in combination with -synteny; size of a maximum common substring (MCS) for adjacency matches; <br> the longer this value becomes the longer syntenic regions need to be in order to be detected </td></tr>
<tr><td>alpha=</td><td>0.5</td><td>weight of adjacencies vs. sequence similarity</td></tr>
</table>

<h3>Using several machines</h3>
If you want to involve multiple machines or separate a Proteinortho run into smaller chunks, use the <i>-steps=</i> option.
First, run <code>proteinortho5.pl -steps=1 ...</code> to generate the indices.
Then you can run <code>proteinortho5.pl -steps=2 -startat=X -stopat=Y ...</code> to run small chunks separately.
Instead of X and Y numbers must be set representing the file number to be calculated. X and Y can be equal.
E.g. you have four fasta files, the you can to run <br>
<code>
proteinortho5.pl -steps=2 -startat=0 -stopat=0 ... <br>
proteinortho5.pl -steps=2 -startat=1 -stopat=1 ... <br>
proteinortho5.pl -steps=2 -startat=2 -stopat=2 ... <br>
proteinortho5.pl -steps=2 -startat=3 -stopat=3 ... <br>
</code>
seperateltly on different machines (within the same shared working directory).
After all step 2 runs are done, <code>proteinortho5.pl -steps=3 ...</code> can be executed to perform the clustering and merge all calculations.
</body>

</html>
